### PR TITLE
Beacon api + validator electra

### PR DESCRIPTION
**Pull request by realbigsean**
## Issue Addressed

- implements this block production change https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/validator.md#deposits

### Actual endpoint changes
- [x] `GET /eth/v2/validator/aggregate_attestation` - this now requires `committee_index` as a query param

### Just the types change
We may need to filter for the `consensus_version_header` in requests, add it to responses, and wrap responses in `ForkVersionedData`. I don't think any of these are required to a LH  BN<>VC connection so I don't think these need to be implemented until they are spec'd.
- [x] `GET /eth/v1/beacon/pool/attestations`
- [x] `POST /eth/v1/beacon/pool/attestations`
- [x] `GET /eth/v1/beacon/pool/attester_slashings`
- [x] `POST /eth/v1/beacon/pool/attester_slashings/`
- [x] `GET /eth/v1/beacon/blocks/{block_id}/attestations`
- [x] `POST /eth/v1/validator/aggregate_and_proofs`

### CODE REVIEW

**Comment by realbigsean** [comment]
spec pr https://github.com/ethereum/beacon-APIs/pull/447

**Review by dapplion — APPROVED** [review]
Straightforward changes, they match the route as spec'ed here https://github.com/ethereum/beacon-APIs/pull/447

**Comment by realbigsean** [comment]
I need to make the fork versioned updates that have been spec'd now: https://github.com/ethereum/beacon-APIs/pull/448

File: beacon_node/http_api/src/lib.rs [hunk 1]
```diff
@@ -1807,20 +1834,40 @@ pub fn serve<T: BeaconChainTypes>(
...
+                        .or_else(|| {
+                            attestations
+                                .first()
+                                .map(|att| att.data().slot)
+                                .or_else(|| chain.slot_clock.now())
+                        })
+                        .ok_or(warp_utils::reject::custom_server_error(
+                            "unable to read slot clock".to_string(),
+                        ))?;
+                    let fork_name = chain.spec.fork_name_at_slot::<T::EthSpec>(slot);
```
Comment (dapplion) [inline beacon_node/http_api/src/lib.rs]: What if someone queries the pool at the fork boundary? There may be attestations from both forks

File: beacon_node/http_api/src/lib.rs [hunk 2]
```diff
@@ -1857,15 +1904,29 @@ pub fn serve<T: BeaconChainTypes>(
...
+                task_spawner.blocking_response_task(Priority::P1, move || {
+                    let slashings = chain.op_pool.get_all_attester_slashings();
+                    let slot = slashings
+                        .first()
+                        .map(|slashing| slashing.attestation_1().data().slot)
+                        .or_else(|| chain.slot_clock.now())
+                        .ok_or(warp_utils::reject::custom_server_error(
+                            "unable to read slot clock".to_string(),
+                        ))?;
+                    let fork_name = chain.spec.fork_name_at_slot::<T::EthSpec>(slot);
```
Comment (dapplion) [inline beacon_node/http_api/src/lib.rs]: Not a reliable way to get the attester slashing fork. One can have an attester slashing electra with an attestation data from deneb. Instead we should implement `fork_name_unchecked` in `slashing`

File: beacon_node/http_api/src/lib.rs [hunk 3]
```diff
@@ -3184,29 +3245,45 @@ pub fn serve<T: BeaconChainTypes>(
...
+            |endpoint_version: EndpointVersion,
+             query: api_types::ValidatorAggregateAttestationQuery,
              not_synced_filter: Result<(), Rejection>,
              task_spawner: TaskSpawner<T::EthSpec>,
              chain: Arc<BeaconChain<T>>| {
                 task_spawner.blocking_json_task(Priority::P0, move || {
                     not_synced_filter?;
-                    chain
-                        .get_pre_electra_aggregated_attestation_by_slot_and_root(
+                    let res = if endpoint_version == V2 {
```
Comment (dapplion) [inline beacon_node/http_api/src/lib.rs:3276]: should `match`?

File: beacon_node/http_api/src/lib.rs (hunk 2 above)
Comment (pawanjay176) [reply beacon_node/http_api/src/lib.rs]: Why aren't we always calculating the fork based on `slot_clock.now()` here?

File: consensus/types/src/attestation.rs [hunk 4]
```diff
@@ -487,6 +487,46 @@ impl<'a, E: EthSpec> From<AttestationRefOnDisk<'a, E>> for AttestationRef<'a, E>
...
+        } else {
+            let attestation: AttestationBase<E> =
+                serde_json::from_value(value).map_err(serde::de::Error::custom)?;
+            Ok(Attestation::Base(attestation))
+        }
+    }
+}
+
+impl<E: EthSpec> ForkVersionDeserialize for Vec<Attestation<E>> {
+    fn deserialize_by_fork<'de, D: serde::Deserializer<'de>>(
```
Comment (pawanjay176) [inline consensus/types/src/attestation.rs:514]: Can we not have a `Vec<Attestation>` with both variants?

File: common/eth2/src/lib.rs [hunk 5]
```diff
@@ -395,6 +408,34 @@ impl BeaconNodeHttpClient {
...
+        body: &T,
+        timeout: Option<Duration>,
+        fork: ForkName,
+    ) -> Result<Response, Error> {
+        let mut builder = self.client.post(url);
+        if let Some(timeout) = timeout {
+            builder = builder.timeout(timeout);
+        }
+
+        let serialized_body = serde_json::to_vec(body).map_err(Error::InvalidJson)?;
```
Comment (pawanjay176) [inline common/eth2/src/lib.rs]: I didn't get why we are doing it this way for attester slashings. Why not just post the json with the content header?

File: beacon_node/http_api/src/lib.rs (hunk 1 above)
Comment (realbigsean) [reply beacon_node/http_api/src/lib.rs]: that's a great point.. looks like we'll have to convert to the new format for all of these attestations

Seems to mean Pawan's idea is the best way to approach it: https://github.com/sigp/lighthouse/pull/5744#discussion_r1672185892

File: beacon_node/http_api/src/lib.rs (hunk 2 above)
Comment (realbigsean) [reply beacon_node/http_api/src/lib.rs]: Good catch

> Why aren't we always calculating the fork based on slot_clock.now() here?

I think this makes sense considering return types could be mixed for a given `Eth-Consensus-Version` as pointed out here  https://github.com/sigp/lighthouse/pull/5744#discussion_r1671810990

File: consensus/types/src/attestation.rs (hunk 4 above)
Comment (realbigsean) [reply consensus/types/src/attestation.rs:514]: I think with these updates no: 386aacda2ab6032ed84fbea7b7d847b0cb5a86d6

This traits for when JSON wraps a data structure with a fork name like this:
```
{
  "version": "phase0",
  "execution_optimistic": false,
  "finalized": false,
  "data": [
    {
      "aggregation_bits": "0x01",
      "data": {
        "slot": "1",
        "index": "1",
        "beacon_block_root": "0xcf8e0d4e9587369b2301d0790347320302cc0943d5a1884560367e8208d920f2",
        "source": {
          "epoch": "1",
          "root": "0xcf8e0d4e9587369b2301d0790347320302cc0943d5a1884560367e8208d920f2"
        },
        "target": {
          "epoch": "1",
          "root": "0xcf8e0d4e9587369b2301d0790347320302cc0943d5a1884560367e8208d920f2"
        }
      },
      "signature": "0x1b66ac1fb663c9bc59509846d6ec05345bd908eda73e670af888da41af171505cc411d61252fb6cb3fa0017b679f8bb2305b26a285fa2737f175668d0dff91cc1b66ac1fb663c9bc59509846d6ec05345bd908eda73e670af888da41af171505"
    }
  ]
}
```

The fork name's meant to be used to deserialize the struct, so i think a consistent fork through the list is implied. More specifically, this is used for when we are displaying attestations from a specific block, which should be uniform, or from the op pool, which should be uniform with the changes listed

File: common/eth2/src/lib.rs (hunk 5 above)
Comment (realbigsean) [reply common/eth2/src/lib.rs]: i sort of just went with what was there already. I actually think this was just included for testing this PR and should have been reverted https://github.com/sigp/lighthouse/pull/4575

File: beacon_node/http_api/src/lib.rs (hunk 2 above)
Comment (realbigsean) [reply beacon_node/http_api/src/lib.rs]: raised a spec issue as well https://github.com/ethereum/beacon-APIs/issues/458

File: common/eth2/src/lib.rs (hunk 5 above)
Comment (realbigsean) [reply common/eth2/src/lib.rs]: updated 71a2eadc467040a4a2996177e557673e5dc104da

File: beacon_node/http_api/src/lib.rs (hunk 3 above)
Comment (realbigsean) [reply beacon_node/http_api/src/lib.rs:3276]: `V2` is just a struct wrapper around a `u64` not an enum

```
pub const V2: EndpointVersion = EndpointVersion(2);

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct EndpointVersion(pub u64);
```

**Comment by realbigsean** [comment]
Thanks for the reviews! Updates have been made

**Review by pawanjay176 — APPROVED** [review]

**Comment by realbigsean** [comment]
In the issue I made on the spec repo, sounds like other teams just drop messages from the op pool that are from the old fork. How do we feel about going that route instead? Might be a little more straightforward for an end user because those message won't actually make it into a block.

**Comment by realbigsean** [comment]
@mergify queue

**Pull request merged**